src/syncer.ts

```diff
@@ -0,0 +1,64 @@
+import type Account from "./account";
+import ElectrumWS, { GetHistoryResponse } from "./electrum";
+
+const GAP_LIMIT = 20;
+
+export async function sync(account: Account, electrum: ElectrumWS): Promise<{ 
+  lastUsed: { internal: number, external: number }, 
+  historyTxsId: Set<string>, 
+  heightsSet: Set<number>,
+  txidHeight: Map<string, number | undefined> 
+}> {
+
+  let historyTxsId: Set<string> = new Set();
+  let heightsSet: Set<number> = new Set();
+  let txidHeight: Map<string, number | undefined> = new Map();
+
+  let lastUsed = { internal: 0, external: 0 };
+  const walletChains = [0, 1];
+  for (const i of walletChains) {
+    const isInternal = i === 1;
+    let batchCount = 0;
+
+    while (true) {
+      const batch = account.deriveBatch(batchCount, GAP_LIMIT, isInternal);
+      const histories = await electrum.batchScriptGetHistory(batch);
+
+      let max = histories
+        .map((v, i) => v.length > 0 ? i : -1)
+        .reduce((a, b) => Math.max(a, b));
+      if (max >= 0) {
+        if (isInternal) {
+          lastUsed.internal = max + batchCount * GAP_LIMIT;
+        } else {
+          lastUsed.external = max + batchCount * GAP_LIMIT;
+        }
+      }
+
+
+      let flattened: GetHistoryResponse[] = histories.flat();
+      console.log(`${i}/batch(${batchCount}) ${flattened.length}`);
+
+      if (flattened.length === 0) {
```

Comment:
this is the interesting trick: we stop looping when the flattened arrays contain 0 elements ie. an empty gap limit